nasa/cumulus
# Jk/cumulus 3319

**Summary:** Summary of changes

Addresses [CUMULUS-3319](https://bugs.earthdata.nasa.gov/browse/CUMULUS-3319)

This PR addresses a situation that we found in testing where a postgres schema requirement (file_name and bucket cannot be null) that was not present in the upstream api schema validation on API writes.  As a result the file would fail to be written, the granule would be updated to failed (as it ALSO does with the framework writes) and then return 200
, which would mask the failure/failure action in the return information (users would have to notice the granule was failed). 

Updating the schema is sufficient to prevent the situation as described above, however this creates downstream effects in testing as well as complexity in the message write chain, and raises questions regarding at which point the granule should be marked failed on API write fail. 

As a result the writeGranules logic was updated to throw if the write is an API write and errors coming back to the user were updated to provide context, however the schema/refactor logic has not (yet) been changed.    This should not change behavior other than the API returns a 400 and has error information. 

## Validation 

Reviewers wishing to see the change in action can: 

On a deployed stack from main:
  * Create required upstream collections, etc and then attempt to PUT a granule (or PATCH ) in such a way that the DB write results in the intended stored object to be missing the fields and view the result/granule state.  
   * Build and update the API with changes from this branch, upload the packaged API lambda (either via deploy or just using AWS cli to update the single lambda) and repeat the prior step. 

## Changes

* Updated API granule write logic to cause postgres schema/db write failures on an individual granule file write to result in a thrown error/400 return instead of a 200 return and a 'silent' update of the granule to failed status.
  * Update api/lib/_writeGranule/_writeGranulefiles logic to allow for schema failures on individual granule writes via an optional method parameter in _writeGranules, and an update to the API granule write calls. 


Comments welcome. 

## PR Checklist
